File: built-in plugins dsp/mb-filter-notch.cpp
/**
 * MB Notch Filter
 * Category : effect
 * Type     : eq
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Precision notch filter for removing specific frequencies
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_FILTER_NOTCH_H
#define MB_FILTER_NOTCH_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbFilterNotch : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-filter-notch";
    static constexpr const char* PLUGIN_NAME    = "MB Notch Filter";
    static constexpr const char* PLUGIN_TYPE    = "eq";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float freq = 1000f;  // range [20, 20000]
    float q = 10f;  // range [1, 50]
    float depth = -60f;  // range [-80, 0]
    };

    MbFilterNotch() = default;
    ~MbFilterNotch() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.freq = std::clamp(params.freq, 20f, 20000f);
        params.q = std::clamp(params.q, 1f, 50f);
        params.depth = std::clamp(params.depth, -80f, 0f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Notch Filter
        return input;
    }
};

#endif // MB_FILTER_NOTCH_H

File: maxbooster7.5/built-in plugins dsp/mb-filter-notch.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbFilterNotchPlugin: PluginDefinition = {
  id: "mb-filter-notch",
  slug: "mb-filter-notch",
  name: "MB Notch Filter",
  category: "effect",
  type: "eq" as any,
  version: "1.0.0",
  description: "Precision notch filter for removing specific frequencies",
  author: "Max Booster",
  grade: "A",
  parameters: [
    {
      id: "freq",
      name: "Frequency",
      type: "float",
      defaultValue: 1000,
      minValue: 20,
      maxValue: 20000,
      automatable: true,
    },
    {
      id: "q",
      name: "Q",
      type: "float",
      defaultValue: 10,
      minValue: 1,
      maxValue: 50,
      automatable: true,
    },
    {
      id: "depth",
      name: "Depth",
      type: "float",
      defaultValue: -60,
      minValue: -80,
      maxValue: 0,
      automatable: true,
    },
  ],
  defaultPreset: { freq: 1000, q: 10, depth: -60 },
};

export default MbFilterNotchPlugin;
